sui_programmability/examples/games_with_chance/sources/satoshi_flip.move

```diff
@@ -0,0 +1,268 @@
+    // min_bet must me lower or equal to max_bet
+    const EMinBetTooHigh: u64 = 4;
+    const EGameAlreadyEnded: u64 = 5; 
+    const ENotEnoughEpochsPassedToCancel: u64 = 6;
+    const EGameNotEnded: u64 = 7;
+    const ENotAllowedToEndGame: u64 = 8;
+    // structs
+
+    // use separate structs for bet and bank so we can extract the coins later
+    struct BetData has store {
+        stake: Balance<SUI>,
+        guess: u8,
+    }
+
+    struct BankData has store {
+        bank_balance: Balance<SUI>,
+        min_bet: u64,
+        max_bet: u64
+    }
+
+    struct Outcome has store {
+        secret: vector<u8>,
+        guess: u8,
+        playerB_won: bool
+    }
+
+    struct Game has key {
+        id: UID,
+        epoch: u64,
+        hashed_secret: Sha3256Digest,
+        playerA: address , // 1st player,
+        playerB: Option<address>, // 2nd player
+        bank_data: Option<BankData>,
+        bet_data: Option<BetData>,
+        outcome: Option<Outcome>
+    }
+
+    // fun!
+
+    // this is invoked by playerA
+    public entry fun start_game(hash: vector<u8>, banker_coin: Coin<SUI>, max_bet: u64, min_bet: u64, ctx: &mut TxContext) {
+        assert!(max_bet >= min_bet, EMinBetTooHigh);
+        let bank_data = BankData {
+            bank_balance: coin::into_balance(banker_coin),
+            min_bet,
+            max_bet
+        };
+
+        let new_game = Game {
+            id: object::new(ctx),
+            epoch: tx_context::epoch(ctx),
+            hashed_secret: digest::sha3_256_digest_from_bytes(hash),
+            playerA: tx_context::sender(ctx),
+            playerB: option::none(),
+            bank_data: option::none(),
+            bet_data: option::none(),
+            outcome: option::none()
+
+        };
+
+        option::fill(&mut new_game.bank_data, bank_data);
+
+        transfer::share_object(new_game); // make it a shared object for now
+    }
+
+    // accessors for playerB to check
+
+    public fun playerA(game: &Game): address {
+        game.playerA
+    }
+
+    public fun max_bet(game: &Game): u64 {
+        let bank_data = option::borrow(&game.bank_data);
+        bank_data.max_bet
+    }
+
+    public fun min_bet(game: &Game): u64 {
+        let bank_data = option::borrow(&game.bank_data);
+        bank_data.min_bet
+    }
+    
+    // these are to view a finished game's result
+    public fun is_playerB_winner(game: &mut Game): bool {
+        let game_outcome = option::borrow(&game.outcome);
+        game_outcome.playerB_won
+    }
+
+    public fun secret(game: &mut Game): vector<u8> {
+        let game_outcome = option::borrow(&game.outcome);
+        game_outcome.secret
+    }
+
+    public fun guess(game: &mut Game): u8 {
+        let game_outcome = option::borrow(&game.outcome);
+        game_outcome.guess
+    }
+
+    // this is invoked by playerB
+    public entry fun bet(game: &mut Game, guess: u8, stake_coin: Coin<SUI>, stake_amount: u64, ctx: &mut TxContext) {
```

Comment:
Iiuc this is not safe! We need to assert that players B bet happened on the allowable period, otherwise they can just bet right before expiration and player A doesn't have enough time to reveal. See this combines with my other message.